nilearn/decoding/tests/test_space_net.py

```diff
@@ -28,50 +29,51 @@
 logistic_path_scores = partial(path_scores, is_classif=True)
 squared_loss_path_scores = partial(path_scores, is_classif=False)
 
-# Data used in almost all tests
 from .test_same_api import to_niimgs
 
-size = 4
-from .simulate_graph_net_data import create_graph_net_simulation_data
+IS_CLASSIF = [True, False]
 
-X_, y, w, mask = create_graph_net_simulation_data(
-    snr=1.0, n_samples=10, size=size, n_points=5, random_state=42
-)
-X, mask = to_niimgs(X_, [size] * 3)
+PENALTY = ["graph-net", "tv-l1"]
 
 
-def test_space_net_alpha_grid(n_samples=4, n_features=3):
+@pytest.mark.parametrize("is_classif", IS_CLASSIF)
+@pytest.mark.parametrize("l1_ratio", [0.5, 1.0])
+@pytest.mark.parametrize("n_alphas", range(1, 10))
+def test_space_net_alpha_grid(
+    is_classif, l1_ratio, n_alphas, n_samples=4, n_features=3
+):
     rng = check_random_state(42)
     X = rng.randn(n_samples, n_features)
     y = np.arange(n_samples)
 
-    for l1_ratio, is_classif in itertools.product([0.5, 1.0], [True, False]):
-        alpha_max = np.max(np.abs(np.dot(X.T, y))) / l1_ratio
-        np.testing.assert_almost_equal(
+    alpha_max = np.max(np.abs(np.dot(X.T, y))) / l1_ratio
+
+    if n_alphas == 1:
+        assert_almost_equal(
             _space_net_alpha_grid(
-                X, y, n_alphas=1, l1_ratio=l1_ratio, logistic=is_classif
+                X, y, n_alphas=n_alphas, l1_ratio=l1_ratio, logistic=is_classif
             ),
             alpha_max,
         )
 
-    for l1_ratio, is_classif in itertools.product([0.5, 1.0], [True, False]):
-        alpha_max = np.max(np.abs(np.dot(X.T, y))) / l1_ratio
-        for n_alphas in range(1, 10):
-            alphas = _space_net_alpha_grid(
-                X, y, n_alphas=n_alphas, l1_ratio=l1_ratio, logistic=is_classif
-            )
-            np.testing.assert_almost_equal(alphas.max(), alpha_max)
-            np.testing.assert_almost_equal(n_alphas, len(alphas))
+    alphas = _space_net_alpha_grid(
+        X, y, n_alphas=n_alphas, l1_ratio=l1_ratio, logistic=is_classif
+    )
+
+    assert_almost_equal(alphas.max(), alpha_max)
+    assert_almost_equal(n_alphas, len(alphas))
 
 
 def test_space_net_alpha_grid_same_as_sk():
+    # for old sklearn versions ???
```

Comment:
Why ?

Reply:
ha!
I put the comment but forgot to get back to it.
My IDE was flagging this import because it could not resolve it:
`from sklearn.linear_model.coordinate_descent import _alpha_grid`

I assumed that this was because this private function was only there for some versions of sklearn and that the test was to be skipped if the import failed.

Turns out the function still exists but in a private module:

https://github.com/scikit-learn/scikit-learn/blob/1ac10736135a62b0dce02c2529f73cd049c31296/sklearn/linear_model/_coordinate_descent.py#L86

So sounds better to change the import: though I would like to know why this test was set up that way in the first place.

Reply:
OK looks like the oldest version of sklearn we support has this function, so I am going to extract the import and remove the try / catch block.